extensions/standard-processors/processors/AttributesToJSON.cpp

```diff
@@ -0,0 +1,174 @@
+                      "Writing to flowfile content will overwrite any existing flowfile content.")
+    ->isRequired(true)
+    ->withDefaultValue<std::string>(toString(WriteDestination::FLOWFILE_ATTRIBUTE))
+    ->withAllowableValues<std::string>(WriteDestination::values())
+    ->build());
+
+const core::Property AttributesToJSON::IncludeCoreAttributes(
+  core::PropertyBuilder::createProperty("Include Core Attributes")
+    ->withDescription("Determines if the FlowFile core attributes which are contained in every FlowFile should be included in the final JSON value generated.")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(true)
+    ->build());
+
+const core::Property AttributesToJSON::NullValue(
+  core::PropertyBuilder::createProperty("Null Value")
+    ->withDescription("If true a non existing or empty attribute will be NULL in the resulting JSON. If false an empty string will be placed in the JSON.")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(false)
+    ->build());
+
+core::Relationship AttributesToJSON::Success("success", "All FlowFiles received are routed to success");
+
+void AttributesToJSON::initialize() {
+  setSupportedProperties({
+    AttributesList,
+    AttributesRegularExpression,
+    Destination,
+    IncludeCoreAttributes,
+    NullValue
+  });
+  setSupportedRelationships({Success});
+}
+
+void AttributesToJSON::onSchedule(core::ProcessContext* context, core::ProcessSessionFactory* /*sessionFactory*/) {
+  std::string value;
+  if (context->getProperty(AttributesList.getName(), value) && !value.empty()) {
+    attribute_list_ = utils::StringUtils::splitAndTrimRemovingEmpty(value, ",");
+  }
+  if (context->getProperty(AttributesRegularExpression.getName(), value) && !value.empty()) {
+    attributes_regular_expression_ = std::regex(value);
+  }
+  write_destination_ = WriteDestination::parse(utils::parsePropertyWithAllowableValuesOrThrow(*context, Destination.getName(), WriteDestination::values()).c_str());
+  context->getProperty(IncludeCoreAttributes.getName(), include_core_attributes_);
+  context->getProperty(NullValue.getName(), null_value_);
+}
+
+bool AttributesToJSON::isCoreAttributeToBeFiltered(const std::string& attribute) const {
+  return !include_core_attributes_ && core_attributes_.find(attribute) != core_attributes_.end();
+}
+
+std::unordered_set<std::string> AttributesToJSON::getAttributesToBeWritten(const std::map<std::string, std::string>& flowfile_attributes) const {
+  std::unordered_set<std::string> attributes;
+
+  for (const auto& attribute : attribute_list_) {
+    if (!isCoreAttributeToBeFiltered(attribute)) {
+      attributes.insert(attribute);
+    }
+  }
+
+  if (attributes_regular_expression_) {
+    for (const auto& [key, value] : flowfile_attributes) {
+      if (!isCoreAttributeToBeFiltered(key) && std::regex_match(key, attributes_regular_expression_.value())) {
```

Comment:
as I see, in the nifi implementation the regex based match does not check if it is a core attribute or not

Reply:
That's true, but should we follow that functionality? It's not really intuitive and seems more like a bug, that when we set the `Include Core Attributes` to false it does not work the same way if the attribute matches the `Attributes List` or the `Attributes Regular Expression` property. IMO if the `Include Core Attributes` is set, then those attributes should not be included.

Reply:
I feel like excluding the core attributes only makes sense when we specify neither the list nor the regex, if the user explicitly requests the attribute by name or regex, I think we should not exclude core ones

Reply:
True, I checked with the NiFi team and they said that the `Include Core Attributes` should only control whether the attributes are included by default, the regex or attribute list should override this.

Reply:
Updated in 16ebbb4cc0bf6256200ee79c755133dec8ae09e2. Thanks for pointing out these discrepancies!